File: ios/Sources/GliaSdkPlugin/GliaSdk.swift
import Capacitor
import Foundation
import GliaWidgets

@objc public class GliaSdk: NSObject {
    
    private var entryWidget: EntryWidget?
    private var authentication: Glia.Authentication?
    
    @objc public func configure(_ call: CAPPluginCall) {
        
        guard
            let siteId = call.getString("siteId"),
            siteId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
        else {
            call.reject("'siteId' is missed or invalid.")
            return
        }
        
        guard
            let apiKey = call.getObject("apiKey"),
            let siteApiKeyId = apiKey["id"] as? String,
            let siteApiKeySecret = apiKey["secret"] as? String
        else {
            call.reject("'apiKey' is missed or invalid.")
            return
        }
        
        guard
            let rawValue = call.getString("region"),
            let region = Environment(rawValue: rawValue)
        else {
            call.reject("'region' is missed or invalid.")
            return
        }
        
        let queueIds = call.getArray("queueIds", []) as? [String]
        
        let companyName = call.getString("companyName") ?? ""
        
        let overrideLocale = call.getString("overrideLocale")
        
        let rawUiConfig = call.getString("uiUnifiedConfig")
        
        var uiConfig: RemoteConfiguration?
        if let jsonData = rawUiConfig?.data(using: .utf8) {
            do {
                uiConfig = try JSONDecoder().decode(RemoteConfiguration.self, from: jsonData)
            } catch {
                debugPrint("Serialization failed with error='\(error)'.")
            }
        }
        
        var visitorContext: Configuration.VisitorContext?
        if let visitorContextAssetId = call.getString("visitorContextAssetId") {
            if let assetId = UUID(uuidString: visitorContextAssetId) {
                visitorContext = .init(assetId: assetId)
            } else {
                call.reject("'visitorContextAssetId' is invalid.")
            }
        }
        
        var features: Features = .all
        if let enableBubbleInsideApp = call.getBool("enableBubbleInsideApp"),
           !enableBubbleInsideApp {
            features.remove(.bubbleView)
        }
        
        let isWhiteLabelApp = call.getBool("isWhiteLabelApp") ?? false
        
        let suppressPushNotificationsPermissionRequestDuringAuthentication = call.getBool("suppressPushNotificationsPermissionRequestDuringAuthentication") ?? false
        
        DispatchQueue.main.async {
            do {
                try Glia.sharedInstance.configure(
                    with: Configuration(
                        authorizationMethod: .siteApiKey(
                            id: siteApiKeyId, secret: siteApiKeySecret),
                        environment: region,
                        site: siteId,
                        visitorContext: visitorContext,
                        isWhiteLabelApp: isWhiteLabelApp,
                        companyName: companyName,
                        manualLocaleOverride: overrideLocale,
                        suppressPushNotificationsPermissionRequestDuringAuthentication: suppressPushNotificationsPermissionRequestDuringAuthentication
                    ),
                    uiConfig: uiConfig,
                    features: features
                ) {  [weak self] (result: Result<Void, Error>) in
                    switch result {
                    case .success:
                        do {
                            self?.entryWidget = try Glia.sharedInstance.getEntryWidget(queueIds: queueIds ?? [])
                            call.resolve()
                        } catch {
                            call.reject("Error occured='\(error)'.")
                        }
                        
                    case .failure(let error):
                        call.reject("Error occured='\(error)'.")
                    }
                }
            } catch {
                call.reject("Error occured='\(error)'.")
            }
        }
    }
    
    @objc public func showEntryWidget(_ call: CAPPluginCall) {
        let queueIds = call.getArray("queueIds") as? [String]
        entryWidget = try Glia.sharedInstance.getEntryWidget(queueIds: queueIds ?? [])
        
        DispatchQueue.main.async { [weak entryWidget] in
            guard let topViewController = UIApplication.topViewController() else {
                call.reject("Could not find view controller for presentatio.")
                return
            }
            
            entryWidget?.show(in: topViewController)
            call.resolve()
        }
    }
    
    @objc public func startChat(_ call: CAPPluginCall) {
        let queueIds = call.getArray("queueIds") as? [String]
        DispatchQueue.main.async {
            do {
                let launcher = try Glia.sharedInstance.getEngagementLauncher(queueIds: queueIds ?? [])
                try launcher.startChat()
                call.resolve()
            } catch {
                call.reject("Could not start chat engagement. Error='\(error)'.")
            }
        }
    }
    
    @objc public func startAudio(_ call: CAPPluginCall) {
        let queueIds = call.getArray("queueIds") as? [String]
        
        DispatchQueue.main.async {
            do {
                let launcher = try Glia.sharedInstance.getEngagementLauncher(queueIds: queueIds ?? [])
                try launcher.startAudioCall()
                call.resolve()
            } catch {
                call.reject("Could not start audio engagement. Error='\(error)'.")
            }
        }
    }
    
    @objc public func startVideo(_ call: CAPPluginCall) {
        let queueIds = call.getArray("queueIds") as? [String]
        DispatchQueue.main.async {
            do {
                let launcher = try Glia.sharedInstance.getEngagementLauncher(queueIds: queueIds ?? [])
                try launcher.startVideoCall()
                call.resolve()
            } catch {
                call.reject("Could not start video engagement. Error'\(error)'.")
            }
        }
    }
    
    @objc public func clearVisitorSession(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Glia.sharedInstance.clearVisitorSession({ [call] (result: Result<Void, Error>) in
                switch result {
                case .success:
                    call.resolve()
                case .failure(let error):
                    call.reject("Could not clear visitor session. Error='\(error)'.")
                }
            })
        }
    }
    
    @objc public func getQueues(_ call: CAPPluginCall) {
        Glia.sharedInstance.getQueues { [call] (result: Result<[Queue], Error>) in
            switch result {
            case .success(let queues):
                call.resolve(
                    queues.reduce(into: [String: Any]()) { _result, queue in
                        _result[queue.id] = [
                            "name": queue.name,
                            "is_default": queue.isDefault,
                            "status": queue.status.stringValue,
                            "media": queue.media.map { $0.rawValue },
                        ]
                    }
                )
                
            case .failure(let error):
                call.reject("Could not get queues. Error='\(error)'.")
            }
        }
    }
    
    @objc public func authenticate(_ call: CAPPluginCall) {
        guard let behavior = call.getString("behavior") else {
            call.reject("'behavior' is missed or invalid.")
            return
        }
        
        guard let idToken = call.getString("idToken") else {
            call.reject("'idToken' is missed or invalid.")
            return
        }
        
        let accessToken = call.getString("accessToken")?.trimmingCharacters(in: .whitespacesAndNewlines)
        
        DispatchQueue.main.async {
            do {
                self.authentication = try Glia.sharedInstance.authentication(
                    with: Glia.Authentication.Behavior(rawValue: behavior)
                )
                self.authentication?.authenticate(
                    with: Glia.Authentication.IdToken(idToken),
                    accessToken: accessToken?.isEmpty == true ? nil : accessToken
                ) { result in
                    switch result {
                    case .success:
                        call.resolve()
                    case .failure(let error):
                        call.reject("Could not authenticate. Error='\(error)'.")
                    }
                }
            } catch {
                call.reject("Could not authenticate. Error='\(error)'.")
            }
        }
    }
    
    @objc public func isAuthenticated(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve(["isAuthenticated": self.authentication?.isAuthenticated ?? false])
        }
    }
    
    @objc public func deauthenticate(_ call: CAPPluginCall) {
        let stopPushNotifications = call.getBool("stopPushNotifications") ?? false
        
        guard let authentication else {
            call.resolve()
            return
        }
        
        DispatchQueue.main.async {
            authentication.deauthenticate(shouldStopPushNotifications: stopPushNotifications) { result in
                switch result {
                case .success:
                    call.resolve()
                case .failure(let error):
                    call.reject("Could not deauthenticate. Error='\(error)'.")
                }
            }
        }
    }
    
    @objc public func refreshAuthentication(_ call: CAPPluginCall) {
        guard let idToken = call.getString("idToken") else {
            call.reject("'idToken' is missed or invalid.")
            return
        }
        
        DispatchQueue.main.async {
            self.authentication?.refresh(
                with: Glia.Authentication.IdToken(idToken),
                accessToken: call.getString("accessToken")
            ) { result in
                switch result {
                case .success:
                    call.resolve()
                case .failure(let error):
                    call.reject("Could not refresh authentication. Error='\(error)'.")
                }
            }
        }
    }
    
    @objc public func showVisitorCode(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let viewController = UIApplication.topViewController() else {
                call.reject("Could not find view controller for presentation.")
                return
            }
            Glia.sharedInstance.callVisualizer.showVisitorCodeViewController(from: viewController)
            call.resolve()
        }
    }
    
    @objc public func startSecureMessaging(_ call: CAPPluginCall) {
        let queueIds = call.getArray("queueIds") as? [String]
        DispatchQueue.main.async {
            do {
                let launcher = try Glia.sharedInstance.getEngagementLauncher(queueIds: queueIds ?? [])
                try launcher.startSecureMessaging()
                call.resolve()
            } catch {
                call.reject("Could not start a Secure Messaging flow. Error='\(error)'.")
            }
        }
    }
    
    @objc public func pauseLiveObservation(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Glia.sharedInstance.liveObservation.pause()
            call.resolve()
        }
    }
    
    @objc public func resumeLiveObservation(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Glia.sharedInstance.liveObservation.resume()
            call.resolve()
        }
    }
}

extension Environment {
    
    init?(rawValue: String) {
        
        switch rawValue.lowercased() {
        case "eu":
            self = .europe
        case "us":
            self = .usa
        case "beta":
            self = .beta
        default:
            return nil
        }
    }
}

extension Glia.Authentication.Behavior {
    init(rawValue: String) {
        switch rawValue.lowercased() {
        case "allowedDuringEngagement":
            self = .allowedDuringEngagement
        default:
            self = .forbiddenDuringEngagement
        }
    }
}

File: android/src/main/java/com/glia/widgets/ionic/GliaSdk.java
package com.glia.widgets.ionic;

import android.app.Activity;
import android.util.Log;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.glia.androidsdk.Glia;
import com.glia.widgets.GliaWidgets;
import com.glia.widgets.GliaWidgetsConfig;
import com.glia.widgets.SiteApiKey;
import com.glia.widgets.authentication.Authentication;
import com.glia.widgets.engagement.MediaType;
import com.glia.widgets.entrywidget.EntryWidget;
import com.glia.widgets.launcher.EngagementLauncher;
import com.glia.widgets.queue.Queue;
import com.glia.widgets.visitor.VisitorInfoUpdateRequest;

import org.json.JSONArray;
import org.json.JSONException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class GliaSdk {
    private List<String> configureQueueIds;

    private EntryWidget entryWidget;

    private Authentication authentication;

    private String visitorContextAssetId;

    public String echo(String value) {
        Log.i("Echo", value);
        return value;
    }

    public void configure(PluginCall call, Activity activity) {

        JSObject apiKey = call.getObject("apiKey");
        String siteApiKeyId = apiKey.getString("id");
        String siteApiKeySecret = apiKey.getString("secret");
        String siteId = call.getString("siteId");
        String region = call.getString("region");
        String companyName = call.getString("companyName");
        JSArray queueIds = call.getArray("queueIds", new JSArray());
        String uiUnifiedConfig = call.getString("uiUnifiedConfig");
        String overrideLocale = call.getString("overrideLocale");
        visitorContextAssetId = call.getString("visitorContextAssetId");
        Boolean enableBubbleOutsideApp = call.getBoolean("enableBubbleOutsideApp", true);
        Boolean enableBubbleInsideApp = call.getBoolean("enableBubbleInsideApp", true);

        if (siteApiKeyId == null) {
            call.reject("'siteApiKeyId' is missing.");
            return;
        }

        if (siteApiKeySecret == null) {
            call.reject("'siteApiKeySecret' is missing.");
            return;
        }

        if (siteId == null) {
            call.reject("'siteId' is missing.");
            return;
        }

        if (region == null) {
            call.reject("'region' is missing.");
            return;
        }

        GliaWidgetsConfig.Builder gliaConfigBuilder = new GliaWidgetsConfig.Builder()
                .setSiteApiKey(new SiteApiKey(siteApiKeyId, siteApiKeySecret))
                .setSiteId(siteId)
                .setRegion(region)
                .setCompanyName(companyName == null ? "" : companyName)
                .setUiJsonRemoteConfig(uiUnifiedConfig)
                .setManualLocaleOverride(overrideLocale)
                .enableBubbleOutsideApp(enableBubbleOutsideApp)
                .enableBubbleInsideApp(enableBubbleInsideApp)
                .setContext(activity.getApplicationContext());

        GliaWidgets.init(
                gliaConfigBuilder.build(),
                () -> {
                    configureQueueIds = jsArrayToArrayList(queueIds);

                    call.resolve();
                },
                exception -> {
                    call.reject(exception.getMessage());
                }
        );
    }

    @Deprecated
    public void presentEntryWidget(PluginCall call, Activity activity) {
        if (configureQueueIds == null) {
            call.reject("SDK is not configured.");
            return;
        }
        entryWidget = GliaWidgets.getEntryWidget(configureQueueIds);
        entryWidget.show(activity);
        call.resolve();
    }

    public void showEntryWidget(PluginCall call, Activity activity) {
        JSArray queueIds = call.getArray("queueIds", new JSArray());
        entryWidget = GliaWidgets.getEntryWidget(jsArrayToArrayList(queueIds));
        entryWidget.show(activity);
        call.resolve();
    }

    public void hideEntryWidget(PluginCall call) {
        entryWidget.hide();
        call.resolve();
    }

    public void startChat(PluginCall call, Activity activity) {
        List<String> queueIds = this.configureQueueIds;
        boolean useOptions = call.getBoolean("useOptions", false);
        if (useOptions) {
            queueIds = jsArrayToArrayList(call.getArray("queueIds", new JSArray()));
        }
        try {
            EngagementLauncher engagementLauncher = GliaWidgets.getEngagementLauncher(queueIds);
            String visitorContextAssetId = this.visitorContextAssetId;
            if (visitorContextAssetId != null && !visitorContextAssetId.isEmpty()) {
                engagementLauncher.startChat(activity, visitorContextAssetId);
            } else {
                engagementLauncher.startChat(activity);
            }
            call.resolve();
        } catch (Exception e) {
            call.reject("Could not start chat engagement. Error='" + e.getMessage() + "'");
        }
    }

    public void startAudio(PluginCall call, Activity activity) {
        List<String> queueIds = this.configureQueueIds;
        boolean useOptions = call.getBoolean("useOptions", false);
        if (useOptions) {
            queueIds = jsArrayToArrayList(call.getArray("queueIds", new JSArray()));
        }
        try {
            EngagementLauncher engagementLauncher = GliaWidgets.getEngagementLauncher(queueIds);
            String visitorContextAssetId = this.visitorContextAssetId;
            if (visitorContextAssetId != null && !visitorContextAssetId.isEmpty()) {
                engagementLauncher.startAudioCall(activity, visitorContextAssetId);
            } else {
                engagementLauncher.startAudioCall(activity);
            }
            call.resolve();
        } catch (Exception e) {
            call.reject("Could not start audio engagement. Error='" + e.getMessage() + "'");
        }
    }

    public void startVideo(PluginCall call, Activity activity) {
        List<String> queueIds = this.configureQueueIds;
        boolean useOptions = call.getBoolean("useOptions", false);
        if (useOptions) {
            queueIds = jsArrayToArrayList(call.getArray("queueIds", new JSArray()));
        }
        try {
            EngagementLauncher engagementLauncher = GliaWidgets.getEngagementLauncher(queueIds);
            String visitorContextAssetId = this.visitorContextAssetId;
            if (visitorContextAssetId != null && !visitorContextAssetId.isEmpty()) {
                engagementLauncher.startVideoCall(activity, visitorContextAssetId);
            } else {
                engagementLauncher.startVideoCall(activity);
            }
            call.resolve();
        } catch (Exception e) {
            call.reject("Could not start video engagement. Error='" + e.getMessage() + "'");
        }
    }

    public void startSecureConversation(PluginCall call, Activity activity) {
        List<String> queueIds = this.configureQueueIds;
        try {
            EngagementLauncher engagementLauncher = GliaWidgets.getEngagementLauncher(queueIds);
            String visitorContextAssetId = this.visitorContextAssetId;
            if (visitorContextAssetId != null && !visitorContextAssetId.isEmpty()) {
                engagementLauncher.startSecureMessaging(activity, visitorContextAssetId);
            } else {
                engagementLauncher.startSecureMessaging(activity);
            }
            call.resolve();
        } catch (Exception e) {
            call.reject("Could not start secure conversation. Error='" + e.getMessage() + "'");
        }
    }

    public void startSecureMessaging(PluginCall call, Activity activity) {
        try {
            List<String> queueIds = jsArrayToArrayList(call.getArray("queueIds", new JSArray()));
            EngagementLauncher engagementLauncher = GliaWidgets.getEngagementLauncher(queueIds);
            String visitorContextAssetId = this.visitorContextAssetId;
            if (visitorContextAssetId != null && !visitorContextAssetId.isEmpty()) {
                engagementLauncher.startSecureMessaging(activity, visitorContextAssetId);
            } else {
                engagementLauncher.startSecureMessaging(activity);
            }
            call.resolve();
        } catch (Exception e) {
            call.reject("Could not start secure messaging. Error='" + e.getMessage() + "'");
        }
    }

    public void clearVisitorSession(PluginCall call) {
        GliaWidgets.clearVisitorSession();
        call.resolve();
    }

    @Deprecated
    public void listQueues(PluginCall call) {
        Glia.getQueues((response, exception) -> {
            if (exception != null) {
                String message = exception != null ? exception.getMessage() : "listQueues failed";
                call.reject(message);
                return;
            } else {
                if (response == null) {
                    call.resolve();
                } else {
                    JSObject jsObject = new JSObject();
                    for (int i = 0; i < response.length; i++) {
                        JSObject child = new JSObject();
                        child.put("name", response[i].getName());
                        child.put("is_default", response[i].isDefault());
                        child.put("status", response[i].getState().getStatus());
                        child.put("media", response[i].getState().getMedias());

                        jsObject.put(response[i].getId(), child);
                    }

                    call.resolve(jsObject);
                }
            }
        });
    }

    public void getQueues(PluginCall call) {
        GliaWidgets.getQueues( response -> {
            if (response == null) {
                call.resolve();
            } else {
                JSObject jsObject = new JSObject();
                for (Queue queue : response) {
                    JSObject child = new JSObject();
                    child.put("name", queue.getName());
                    child.put("is_default", queue.isDefault());
                    child.put("status", serializedName(queue.getStatus()));

                    JSONArray mediaTypes = new JSONArray();
                    for (MediaType mediaType : queue.getMedia()) {
                        mediaTypes.put(serializedName(mediaType));
                    }
                    child.put("media", mediaTypes);

                    jsObject.put(queue.getId(), child);
                }

                call.resolve(jsObject);
            }
        }, exception -> {
            call.reject(exception.getMessage());
        });
    }

    public void showVisitorCode(PluginCall call) {
        GliaWidgets.getCallVisualizer().showVisitorCodeDialog();
        call.resolve();
    }

    public void authenticate(PluginCall call) {
        String authRawBehaviorValue = call.getString("behavior");
        String jwtToken = call.getString("idToken");

        if (jwtToken == null) {
            call.reject("'idToken' is missing");
            return;
        }
        if (authRawBehaviorValue == null) {
            call.reject("Authentication behavior is missing.");
            return;
        }

        String accessToken = call.getString("accessToken", null);
        if (accessToken != null && accessToken.isEmpty()) {
            accessToken = null;
        }
        AuthenticationBehavior authBehavior = AuthenticationBehavior.valueOf(authRawBehaviorValue);

        authentication = GliaWidgets.getAuthentication(authBehavior.toSdkNativeType());
        authentication.authenticate(
                jwtToken,
                accessToken,
                call::resolve,
                exception -> call.reject(exception.getMessage())
        );
    }

    public void deauthenticate(PluginCall call) {
        if (authentication == null) {
            call.resolve();
            return;
        }
        authentication.deauthenticate(
                call::resolve,
                exception -> call.reject(exception.getMessage())
        );
    }

    public void isAuthenticated(PluginCall call) {
        call.resolve(new JSObject().put("isAuthenticated", authentication != null && authentication.isAuthenticated()));
    }

    public void refreshAuthentication(PluginCall call) {
        String jwtToken = call.getString("idToken");
        String accessToken = call.getString("accessToken");
        if (jwtToken == null) {
            call.reject("'idToken' is missing");
            return;
        }
        if (accessToken != null && accessToken.isEmpty()) {
            accessToken = null;
        }
        authentication.refresh(
                jwtToken,
                accessToken,
                call::resolve,
                exception -> call.reject(exception.getMessage())
        );
    }

    public void pauseLiveObservation(PluginCall call) {
        GliaWidgets.getLiveObservation().pause();
        call.resolve();
    }

    public void resumeLiveObservation(PluginCall call) {
        GliaWidgets.getLiveObservation().resume();
        call.resolve();
    }

    public void getVisitorInfo(PluginCall call) {
        try {
            GliaWidgets.getVisitorInfo(visitorInfo -> {
                JSObject result = new JSObject();
                result.put("name", visitorInfo.getName());
                result.put("email", visitorInfo.getEmail());
                result.put("phone", visitorInfo.getPhone());
                result.put("note", visitorInfo.getNote());
                result.put("externalId", visitorInfo.getExternalId());

                JSObject customAttributes = new JSObject();
                if (visitorInfo.getCustomAttributes() != null) {
                    for (String key : visitorInfo.getCustomAttributes().keySet()) {
                        customAttributes.put(key, visitorInfo.getCustomAttributes().get(key));
                    }
                }
                result.put("customAttributes", customAttributes);
                result.put("banned", visitorInfo.getBanned());

                call.resolve(result);
            }, exception -> {
                String message = exception.getMessage() != null ? exception.getMessage() : "Could not fetch visitor info.";
                call.reject(message);
            });
        } catch (Exception exception) {
            String message = exception.getMessage() != null ? exception.getMessage() : "Could not fetch visitor info.";
            call.reject(message);
        }
    }

    public void updateVisitorInfo(PluginCall call) {
        try {
            String name = call.getString("name");
            String email = call.getString("email");
            String phone = call.getString("phone");
            String note = call.getString("note");
            String noteUpdateMethod = call.getString("noteUpdateMethod");
            String externalId = call.getString("externalId");
            String customAttrsUpdateMethod = call.getString("customAttributesUpdateMethod");

            JSObject customAttributesObj = call.getObject("customAttributes");
            java.util.Map<String, String> customAttributes = new java.util.HashMap<>();
            if (customAttributesObj != null) {
                Iterator<String> keys = customAttributesObj.keys();
                while (keys.hasNext()) {
                    String key = keys.next();
                    Object value = customAttributesObj.get(key);
                    if (value instanceof String) {
                        customAttributes.put(key, (String) value);
                    }
                }
            }

            VisitorInfoUpdateRequest updateRequest = new VisitorInfoUpdateRequest();
            updateRequest.setName(name);
            updateRequest.setEmail(email);
            updateRequest.setPhone(phone);
            updateRequest.setNote(note);
            if (noteUpdateMethod != null) {
                updateRequest.setNoteUpdateMethod(toNoteUpdateMethod(noteUpdateMethod));
            }
            updateRequest.setExternalId(externalId);
            updateRequest.setCustomAttributes(customAttributes);
            if (customAttrsUpdateMethod != null) {
                updateRequest.setCustomAttrsUpdateMethod(toCustomAttributesUpdateMethod(customAttrsUpdateMethod));
            }

            GliaWidgets.updateVisitorInfo(
                    updateRequest,
                    call::resolve,
                    exception -> {
                        call.reject(exception.getMessage());
                    }
            );
        } catch (Exception exception) {
            call.reject("Could not update visitor info. Error='" + exception.getMessage() + "'");
        }
    }

    public void endEngagement(PluginCall call) {
        GliaWidgets.endEngagement();
        call.resolve();
    }

    public ArrayList<String> jsArrayToArrayList(JSArray jsArray) {
        if (jsArray == null) {
            return new ArrayList<>();
        }
        ArrayList<String> arrayList = new ArrayList<>();

        try {
            for (int i = 0; i < jsArray.length(); i++) {
                arrayList.add(jsArray.getString(i));  // Assumes each element is a String
            }
        } catch (JSONException e) {
            e.printStackTrace();
            // Handle the exception based on your needs, e.g., log the error or notify the user
        }

        return arrayList;
    }

    private String serializedName(Queue.Status status) {
        return switch (status) {
            case OPEN -> "opened";
            case CLOSED -> "closed";
            case FULL -> "full";
            case UNSTAFFED -> "unstaffed";
            default -> "unknown";
        };
    }

    private String serializedName(MediaType mediaType) {
        return switch (mediaType) {
            case TEXT -> "text";
            case AUDIO -> "audio";
            case VIDEO -> "video";
            case PHONE -> "phone";
            case MESSAGING -> "messaging";
            default -> "unknown";
        };
    }

    private VisitorInfoUpdateRequest.NoteUpdateMethod toNoteUpdateMethod(String value) {
        return switch (value.toLowerCase()) {
            case "replace" -> VisitorInfoUpdateRequest.NoteUpdateMethod.REPLACE;
            default -> VisitorInfoUpdateRequest.NoteUpdateMethod.APPEND;
        };
    }

    private VisitorInfoUpdateRequest.CustomAttributesUpdateMethod toCustomAttributesUpdateMethod(String value) {
        return switch (value.toLowerCase()) {
            case "replace" -> VisitorInfoUpdateRequest.CustomAttributesUpdateMethod.REPLACE;
            default -> VisitorInfoUpdateRequest.CustomAttributesUpdateMethod.MERGE;
        };
    }
}